### ISSUE TITLE
new enum type?

### ISSUE THREAD

**Issue by rurban** [perl11/cperl#307]
See https://docs.perl6.org/language/typesystem#index-entry-Enumeration-_Enums-_enum
Basically constant typed arrays (Int or Str), starting with 0, with auto-incrementing values which can be overridden by using an optional pair syntax in the decl.

The current perl6 design is a bit *unfortunate* (i.e. broken), 
as barewords don't return the value, but the name,
and they look now like hashes, not arrays.
`enum Names ( name1 => 1, name2 => 2 );`
`say name1, ' ', name2; # OUTPUT: «name1 name2␤»`
and you need to manually expand it: `say name1.value, ' ', name2.value; # OUTPUT: «1 2␤»`
and not the other way round, as expected. They also say Number, not Int. (i.e. + Num, Rat. Wonder how to increment those).

The old design was correct: https://design.perl6.org/S12.html#Enumerations
That is,
```
     enum E <a b c>;
```
is largely syntactic sugar for:
```
     package E {
        constant a = 0;
        constant b = 1;
        constant c = 2;
    }
```

The cperl plan is:
```
enum E (a b c);
enum E1 (d => 1, e, f);  # but not the perl6 syntax: `enum E (:a(1), b c);` 
enum E2 (g, h => 4, i);  # pairs allowed everywhere 
print a; # OUTPUT: «1»   but not «a»

enum E3 (j => 1.0, k, l); # => error: Invalid enum type Num 1.0
```

**Comment by vendethiel** [comment]
Do you have any idea when P6 changed semantics?

**Comment by rurban** [comment]
No. Looks like a typical driveby-design-change fuckup, as in many other languages. Bjarne Stroustrup was famous to say yes to anyone and everything just to get C++ accepted.
